File: zig/src/fetch.zig
const std = @import("std");

const packfile = @import("packfile.zig");
const utils = @import("utils.zig");
const remote = @import("remote.zig");

pub const FetchResult = struct {
    remote: []const u8,
    refs: std.ArrayList(RefInfo),
};

pub const RefInfo = struct {
    name: []const u8,
    sha: []const u8,
};

const DiscoveredRef = struct {
    sha: []const u8,
    ref: []const u8,
};

pub fn fetchFromRemote(io: std.Io, allocator: std.mem.Allocator, path: []const u8, remote_name_param: ?[]const u8) !FetchResult {
    const git_dir_path = try std.fs.path.join(allocator, &[_][]const u8{ path, ".git" });
    defer allocator.free(git_dir_path);

    const cwd = std.Io.Dir.cwd();
    const git_dir = cwd.openDir(io, git_dir_path, .{}) catch {
        return error.NotAGitRepository;
    };
    git_dir.close(io);

    const remote_name = if (remote_name_param) |r| r else "origin";
    const remote_url_opt = try remote.getRemoteUrl(io, allocator, git_dir_path, remote_name);

    if (remote_url_opt == null) {
        return FetchResult{
            .remote = try allocator.dupe(u8, remote_name),
            .refs = std.ArrayList(RefInfo).initCapacity(allocator, 0) catch unreachable,
        };
    }

    const remote_url = remote_url_opt.?;
    defer allocator.free(remote_url);

    var refs = try discoverRefs(io, allocator, remote_url);
    defer {
        for (refs.items) |ref| {
            allocator.free(ref.sha);
            allocator.free(ref.ref);
        }
        refs.deinit(allocator);
    }

    var result = FetchResult{
        .remote = try allocator.dupe(u8, remote_name),
        .refs = std.ArrayList(RefInfo).initCapacity(allocator, 0) catch unreachable,
    };
    errdefer {
        allocator.free(result.remote);
        for (result.refs.items) |ref_info| {
            allocator.free(ref_info.name);
            allocator.free(ref_info.sha);
        }
        result.refs.deinit(allocator);
    }

    var wants = std.ArrayList([]const u8).initCapacity(allocator, 0) catch unreachable;
    defer {
        for (wants.items) |want| allocator.free(want);
        wants.deinit(allocator);
    }

    var haves = std.ArrayList([]const u8).initCapacity(allocator, 0) catch unreachable;
    defer {
        for (haves.items) |have| allocator.free(have);
        haves.deinit(allocator);
    }

    for (refs.items) |ref| {
        if (std.mem.startsWith(u8, ref.ref, "refs/heads/")) {
            const branch_name = ref.ref["refs/heads/".len..];
            const local_ref_path = try std.fs.path.join(allocator, &[_][]const u8{ git_dir_path, "refs", "heads", branch_name });
            defer allocator.free(local_ref_path);

            if (cwd.access(io, local_ref_path, .{})) |_| {
                const local_sha_with_newline = try cwd.readFileAlloc(io, local_ref_path, allocator, .unlimited);
                defer allocator.free(local_sha_with_newline);

                const local_sha = std.mem.trim(u8, local_sha_with_newline, &std.ascii.whitespace);
                try haves.append(allocator, try allocator.dupe(u8, local_sha));
            } else |err| {
                if (err != error.FileNotFound) {
                    return err;
                }
            }

            try wants.append(allocator, try allocator.dupe(u8, ref.sha));
            try result.refs.append(allocator, .{
                .name = try allocator.dupe(u8, branch_name),
                .sha = try allocator.dupe(u8, ref.sha),
            });
        }
    }

    if (wants.items.len > 0) {
        var objects = try fetchPackfile(io, allocator, remote_url, wants, haves);
        defer {
            for (objects.items) |obj| {
                allocator.free(obj.obj_type);
                allocator.free(obj.sha);
                allocator.free(obj.content);
            }
            objects.deinit(allocator);
        }

        try storeObjects(io, allocator, git_dir_path, objects);
    }

    for (result.refs.items) |ref_info| {
        const remote_ref_path = try std.fs.path.join(allocator, &[_][]const u8{ git_dir_path, "refs", "remotes", remote_name, ref_info.name });
        defer allocator.free(remote_ref_path);

        const remote_ref_dir = try std.fs.path.join(allocator, &[_][]const u8{ git_dir_path, "refs", "remotes", remote_name });
        defer allocator.free(remote_ref_dir);

        try cwd.createDirPath(io, remote_ref_dir);

        const commit_with_newline = try std.fmt.allocPrint(allocator, "{s}\n", .{ref_info.sha});
        defer allocator.free(commit_with_newline);

        try cwd.writeFile(io, .{ .sub_path = remote_ref_path, .data = commit_with_newline });
    }

    return result;
}

fn storeObjects(io: std.Io, allocator: std.mem.Allocator, git_dir_path: []const u8, objects: std.ArrayList(packfile.PackObject)) !void {
    for (objects.items) |obj| {
        _ = try utils.hashObjectBinary(io, allocator, git_dir_path, obj.content, obj.obj_type);
    }
}

fn discoverRefs(io: std.Io, allocator: std.mem.Allocator, remote_url: []const u8) !std.ArrayList(DiscoveredRef) {
    var url_parts = std.mem.splitScalar(u8, remote_url, '/');
    var host: []const u8 = undefined;
    var path: []const u8 = undefined;

    var i: usize = 0;
    while (url_parts.next()) |part| {
        if (i == 0) {
            if (std.mem.startsWith(u8, part, "http://")) {
                host = part["http://".len..];
            } else if (std.mem.startsWith(u8, part, "https://")) {
                host = part["https://".len..];
            } else {
                host = part;
            }
        } else {
            path = part;
        }
        i += 1;
    }

    const full_url = try std.fmt.allocPrint(allocator, "http://{s}/info/refs?service=git-upload-pack", .{host});
    defer allocator.free(full_url);

    var client: std.http.Client = .{ .allocator = allocator, .io = io };
    defer client.deinit();

    const uri = try std.Uri.parse(full_url);
    var request = try client.request(.GET, uri, .{});
    defer request.deinit();

    try request.sendBodiless();

    var redirect_buffer: [4096]u8 = undefined;
    var response = try request.receiveHead(&redirect_buffer);

    if (response.head.status != .ok) {
        return error.FailedToDiscoverRefs;
    }

    var body = std.ArrayList(u8).initCapacity(allocator, 0) catch unreachable;
    defer body.deinit(allocator);

    var transfer_buffer: [8192]u8 = undefined;
    const reader = response.reader(&transfer_buffer);

    try reader.appendRemainingUnlimited(allocator, &body);

    var lines = try packfile.decodePktLines(allocator, body.items);
    defer {
        for (lines.items) |line| allocator.free(line);
        lines.deinit(allocator);
    }

    var refs = std.ArrayList(DiscoveredRef).initCapacity(allocator, 0) catch unreachable;
    errdefer {
        for (refs.items) |ref| {
            allocator.free(ref.sha);
            allocator.free(ref.ref);
        }
        refs.deinit(allocator);
    }

    var started = false;

    for (lines.items) |line| {
        if (std.mem.indexOf(u8, line, "# service=git-upload-pack")) |_| {
            started = true;
            continue;
        }

        if (!started) continue;
        if (line.len == 0) continue;

        const trimmed = std.mem.trim(u8, line, &std.ascii.whitespace);
        if (trimmed.len == 0) continue;

        var parts = std.mem.splitScalar(u8, trimmed, ' ');
        const sha_str = parts.next() orelse continue;
        const ref_str = parts.next() orelse continue;

        if (sha_str.len != 40) continue;

        const is_valid_sha = blk: {
            for (sha_str) |c| {
                if (!(std.ascii.isDigit(c) or (std.ascii.isHex(c)))) break :blk false;
            }
            break :blk true;
        };

        if (!is_valid_sha) continue;

        const null_idx = std.mem.indexOfScalar(u8, ref_str, 0) orelse ref_str.len;
        const ref_name = std.mem.trim(u8, ref_str[0..null_idx], &std.ascii.whitespace);

        try refs.append(allocator, .{
            .sha = try allocator.dupe(u8, sha_str),
            .ref = try allocator.dupe(u8, ref_name),
        });
    }

    return refs;
}

fn buildLsRefsRequest(allocator: std.mem.Allocator) ![]const u8 {
    var lines = std.ArrayList([]const u8).initCapacity(allocator, 0) catch unreachable;
    errdefer {
        for (lines.items) |line| allocator.free(line);
        lines.deinit(allocator);
    }

    const command_pkt = try packfile.encodePktLine(allocator, "command=ls-refs\n");
    try lines.append(allocator, command_pkt);

    const delimiter_pkt = try allocator.dupe(u8, "0001");
    try lines.append(allocator, delimiter_pkt);

    const symrefs_pkt = try packfile.encodePktLine(allocator, "symrefs\n");
    try lines.append(allocator, symrefs_pkt);

    const peel_pkt = try packfile.encodePktLine(allocator, "peel\n");
    try lines.append(allocator, peel_pkt);

    const ref_prefix_pkt = try packfile.encodePktLine(allocator, "ref-prefix refs/heads/\n");
    try lines.append(allocator, ref_prefix_pkt);

    const flush_pkt = try packfile.encodePktLine(allocator, null);
    try lines.append(allocator, flush_pkt);

    var result = std.ArrayList(u8).initCapacity(allocator, 0) catch unreachable;
    for (lines.items) |line| {
        try result.appendSlice(allocator, line);
    }

    return result.toOwnedSlice(allocator);
}

fn extractPackfileFromSideband(allocator: std.mem.Allocator, data: []const u8) ![]const u8 {
    var packfile_data = std.ArrayList(u8).initCapacity(allocator, 0) catch unreachable;
    errdefer packfile_data.deinit(allocator);

    var offset: usize = 0;

    while (offset < data.len) {
        if (offset + 4 > data.len) {
            break;
        }

        const hex_length = data[offset .. offset + 4];
        if (std.mem.eql(u8, hex_length, "0000")) {
            offset += 4;
            continue;
        }

        if (std.mem.eql(u8, hex_length, "0001")) {
            offset += 4;
            continue;
        }

        const length = std.fmt.parseUnsigned(usize, hex_length, 16) catch break;
        if (length == 0 or offset + length > data.len) {
            break;
        }

        const payload = data[offset + 4 .. offset + length];

        if (payload.len > 0) {
            const channel = payload[0];
            if (channel == 1) {
                try packfile_data.appendSlice(allocator, payload[1..]);
            } else if (channel == 3) {
                const error_msg = payload[1..];
                std.debug.print("Git error: {s}\n", .{error_msg});
            }
        }

        offset += length;
    }

    return packfile_data.toOwnedSlice(allocator);
}

fn fetchPackfile(io: std.Io, allocator: std.mem.Allocator, remote_url: []const u8, wants: std.ArrayList([]const u8), haves: std.ArrayList([]const u8)) !std.ArrayList(packfile.PackObject) {
    var url_parts = std.mem.splitScalar(u8, remote_url, '/');
    var host: []const u8 = undefined;

    var i: usize = 0;
    while (url_parts.next()) |part| {
        if (i == 0) {
            if (std.mem.startsWith(u8, part, "http://")) {
                host = part["http://".len..];
            } else if (std.mem.startsWith(u8, part, "https://")) {
                host = part["https://".len..];
            } else {
                host = part;
            }
        }
        i += 1;
    }

    const full_url = try std.fmt.allocPrint(allocator, "http://{s}/git-upload-pack", .{host});
    defer allocator.free(full_url);

    var client: std.http.Client = .{ .allocator = allocator, .io = io };
    defer client.deinit();

    const uri = try std.Uri.parse(full_url);
    var request = try client.request(.POST, uri, .{});
    defer request.deinit();

    const request_body = try buildFetchRequest(allocator, wants, haves);
    defer allocator.free(request_body);

    try request.sendBodyComplete(@constCast(request_body));

    var redirect_buffer: [4096]u8 = undefined;
    var response = try request.receiveHead(&redirect_buffer);

    if (response.head.status != .ok) {
        return error.FailedToFetchPackfile;
    }

    var body = std.ArrayList(u8).initCapacity(allocator, 0) catch unreachable;
    defer body.deinit(allocator);

    var transfer_buffer: [8192]u8 = undefined;
    const reader = response.reader(&transfer_buffer);

    try reader.appendRemainingUnlimited(allocator, &body);

    const packfile_data = try extractPackfileFromSideband(allocator, body.items);
    defer allocator.free(packfile_data);

    if (packfile_data.len == 0) {
        return std.ArrayList(packfile.PackObject).initCapacity(allocator, 0) catch unreachable;
    }

    return try packfile.parsePackfile(allocator, packfile_data);
}

fn buildFetchRequest(allocator: std.mem.Allocator, wants: std.ArrayList([]const u8), haves: std.ArrayList([]const u8)) ![]const u8 {
    var lines = std.ArrayList([]const u8).initCapacity(allocator, 0) catch unreachable;
    errdefer {
        for (lines.items) |line| allocator.free(line);
        lines.deinit(allocator);
    }

    const command_pkt = try packfile.encodePktLine(allocator, "command=fetch\n");
    try lines.append(allocator, command_pkt);

    const delimiter_pkt = try allocator.dupe(u8, "0001");
    try lines.append(allocator, delimiter_pkt);

    for (wants.items) |w| {
        const pkt_line = try packfile.encodePktLine(allocator, try std.fmt.allocPrint(allocator, "want {s}\n", .{w}));
        try lines.append(allocator, pkt_line);
    }

    for (haves.items) |h| {
        const pkt_line = try packfile.encodePktLine(allocator, try std.fmt.allocPrint(allocator, "have {s}\n", .{h}));
        try lines.append(allocator, pkt_line);
    }

    const done_pkt = try packfile.encodePktLine(allocator, "done\n");
    try lines.append(allocator, done_pkt);

    const flush_pkt = try packfile.encodePktLine(allocator, null);
    try lines.append(allocator, flush_pkt);

    var result = std.ArrayList(u8).initCapacity(allocator, 0) catch unreachable;
    for (lines.items) |line| {
        try result.appendSlice(allocator, line);
    }

    return result.toOwnedSlice(allocator);
}

File: swift/Sources/Gitologist/Fetch.swift
import Foundation

struct FetchResult {
	let remote: String
	var refs: [RefInfo]
}

struct RefInfo {
	let name: String
	let sha: String
}

enum FetchError: Error, LocalizedError {
	case notAGitRepository
	case failedToDiscoverRefs(Int, String)
	case failedToFetchPackfile(Int, String)
	case invalidURL

	var errorDescription: String? {
		switch self {
		case .notAGitRepository:
			return "Not a git repository"
		case let .failedToDiscoverRefs(status, text):
			return "Failed to discover refs: \(status) \(text)"
		case let .failedToFetchPackfile(status, text):
			return "Failed to fetch packfile: \(status) \(text)"
		case .invalidURL:
			return "Invalid URL"
		}
	}
}

func fetchFromRemote(at path: String, remote: String? = nil) async throws -> FetchResult {
	let gitDir = URL(fileURLWithPath: path).appendingPathComponent(".git")

	guard FileManager.default.fileExists(atPath: gitDir.path) else {
		throw FetchError.notAGitRepository
	}

	let remoteName = remote ?? "origin"
	guard let remoteUrl = await getRemoteUrl(at: gitDir.path, remoteName: remoteName) else {
		return FetchResult(remote: remoteName, refs: [])
	}

	let refs = try await discoverRefs(remoteUrl: remoteUrl)
	var result = FetchResult(remote: remoteName, refs: [])

	var wants: [String] = []
	var haves: [String] = []

	for ref in refs {
		if ref.ref.hasPrefix("refs/heads/") {
			let branchName = String(ref.ref.dropFirst("refs/heads/".count))
			let localRefPath = gitDir.appendingPathComponent("refs").appendingPathComponent("heads").appendingPathComponent(branchName)

			if FileManager.default.fileExists(atPath: localRefPath.path) {
				let localSha = try String(contentsOf: localRefPath, encoding: .utf8).trimmingCharacters(in: .whitespacesAndNewlines)
				haves.append(localSha)
			}

			wants.append(ref.sha)
			result.refs.append(RefInfo(name: branchName, sha: ref.sha))
		}
	}

	if !wants.isEmpty {
		let objects = try await fetchPackfile(remoteUrl: remoteUrl, wants: wants, haves: haves)
		try await storeObjects(at: gitDir.path, objects: objects)
	}

	for ref in result.refs {
		let remoteRefPath = gitDir
			.appendingPathComponent("refs")
			.appendingPathComponent("remotes")
			.appendingPathComponent(remoteName)
			.appendingPathComponent(ref.name)
		try FileManager.default.createDirectory(at: remoteRefPath.deletingLastPathComponent(), withIntermediateDirectories: true)
		try "\(ref.sha)\n".write(to: remoteRefPath, atomically: true, encoding: .utf8)
	}

	return result
}

func storeObjects(at gitDir: String, objects: [PackObject]) async throws {
	for obj in objects {
		_ = try await hashObject(at: gitDir, data: obj.content, type: obj.type.rawValue)
	}
}

struct DiscoveredRef {
	let sha: String
	let ref: String
}

func discoverRefs(remoteUrl: String) async throws -> [DiscoveredRef] {
	guard let url = URL(string: remoteUrl) else {
		throw FetchError.invalidURL
	}
	var components = URLComponents(url: url.appendingPathComponent("info/refs"), resolvingAgainstBaseURL: true)
	components?.queryItems = [URLQueryItem(name: "service", value: "git-upload-pack")]

	guard let fetchUrl = components?.url else {
		throw FetchError.invalidURL
	}

	var request = URLRequest(url: fetchUrl)
	request.httpMethod = "GET"
	request.setValue("application/x-git-upload-pack-advertisement", forHTTPHeaderField: "Accept")
	request.setValue("version=2", forHTTPHeaderField: "Git-Protocol")

	let (data, response) = try await URLSession.shared.data(for: request)

	guard let httpResponse = response as? HTTPURLResponse else {
		throw FetchError.failedToDiscoverRefs(0, "Invalid response")
	}

	guard httpResponse.statusCode == 200 else {
		throw FetchError.failedToDiscoverRefs(httpResponse.statusCode, HTTPURLResponse.localizedString(forStatusCode: httpResponse.statusCode))
	}

	let lines = decodePktLines(data)

	var refs: [DiscoveredRef] = []
	var started = false

	for line in lines {
		if line.contains("# service=git-upload-pack") {
			started = true
			continue
		}

		if !started {
			continue
		}
		if line.isEmpty {
			continue
		}

		let parts = line.split(separator: " ", maxSplits: 1)
		if parts.count >= 2 {
			let sha = String(parts[0])
			guard sha.count == 40, sha.allSatisfy({ $0.isNumber || ($0.isLetter && $0.isASCII) }) else {
				continue
			}
			let refParts = String(parts[1]).split(separator: "\0")
			guard !refParts.isEmpty else {
				continue
			}
			let ref = String(refParts[0])
			refs.append(DiscoveredRef(sha: sha, ref: ref))
		}
	}

	return refs
}

func fetchPackfile(remoteUrl: String, wants: [String], haves: [String]) async throws -> [PackObject] {
	guard let url = URL(string: remoteUrl) else {
		throw FetchError.invalidURL
	}
	let fetchUrl = url.appendingPathComponent("git-upload-pack")

	let requestBody = buildFetchRequest(wants: wants, haves: haves)

	var request = URLRequest(url: fetchUrl)
	request.httpMethod = "POST"
	request.setValue("application/x-git-upload-pack-request", forHTTPHeaderField: "Content-Type")
	request.setValue("application/x-git-upload-pack-result", forHTTPHeaderField: "Accept")
	request.httpBody = requestBody

	let (data, response) = try await URLSession.shared.data(for: request)

	guard let httpResponse = response as? HTTPURLResponse else {
		throw FetchError.failedToFetchPackfile(0, "Invalid response")
	}

	guard httpResponse.statusCode == 200 else {
		throw FetchError.failedToFetchPackfile(httpResponse.statusCode, HTTPURLResponse.localizedString(forStatusCode: httpResponse.statusCode))
	}

	guard let packfileOffset = findPackfileStart(data) else {
		return []
	}

	let packfileData = data[packfileOffset...]
	return try parsePackfile(Data(packfileData))
}

func buildFetchRequest(wants: [String], haves: [String]) -> Data {
	var lines: [Data] = []

	for want in wants {
		lines.append(encodePktLine("want \(want)"))
	}

	for have in haves {
		lines.append(encodePktLine("have \(have)"))
	}

	lines.append(encodePktLine("done"))
	lines.append(encodePktLine(nil))

	return lines.reduce(Data(), +)
}

func findPackfileStart(_ data: Data) -> Data.Index? {
	let signature = Data("PACK".utf8)
	for i in 0 ..< (data.count - 4) {
		if data[i ..< (i + 4)] == signature {
			return i
		}
	}
	return nil
}
